pkg/generator/generator.go

```diff
@@ -351,52 +415,159 @@ func (g *Generator) renderPkg() error {
 	v := version(g.apiVersion)
 	adn := apiDirName(g.apiVersion)
 	apiDir := filepath.Join(g.projectName, apisDir, adn, v)
-	if err := os.MkdirAll(apiDir, defaultDirFileMode); err != nil {
-		return err
-	}
 	if err := renderAPIFiles(apiDir, groupName(g.apiVersion), v, g.kind); err != nil {
 		return err
 	}
 
 	sDir := filepath.Join(g.projectName, stubDir)
-	if err := os.MkdirAll(sDir, defaultDirFileMode); err != nil {
-		return err
-	}
 	return renderStubFiles(sDir, g.repoPath, g.kind, adn, v)
 }
 
 func renderAPIFiles(apiDir, groupName, version, kind string) error {
 	buf := &bytes.Buffer{}
-	if err := renderAPIDocFile(buf, groupName, version); err != nil {
+	if err := renderGenericFile(buf, "apis/<apiDirName>/<version>/doc.go", apiDocTmpl, tmplData{
+		GroupName: groupName,
+		Version:   version,
+	}); err != nil {
 		return err
 	}
 	if err := writeFileAndPrint(filepath.Join(apiDir, doc), buf.Bytes(), defaultFileMode); err != nil {
 		return err
 	}
 
 	buf = &bytes.Buffer{}
-	if err := renderAPIRegisterFile(buf, kind, groupName, version); err != nil {
+	if err := renderGenericFile(buf, "apis/<apiDirName>/<version>/register.go", apiRegisterTmpl, tmplData{
+		Kind:       kind,
+		KindPlural: toPlural(strings.ToLower(kind)),
+		GroupName:  groupName,
+		Version:    version,
+	}); err != nil {
 		return err
 	}
 	if err := writeFileAndPrint(filepath.Join(apiDir, register), buf.Bytes(), defaultFileMode); err != nil {
 		return err
 	}
 
 	buf = &bytes.Buffer{}
-	if err := renderAPITypesFile(buf, kind, version); err != nil {
+	if err := renderGenericFile(buf, "apis/<apiDirName>/<version>/types.go", apiTypesTmpl, tmplData{
+		Kind:    kind,
+		Version: version,
+	}); err != nil {
 		return err
 	}
 	return writeFileAndPrint(filepath.Join(apiDir, types), buf.Bytes(), defaultFileMode)
 }
 
 func renderStubFiles(stubDir, repoPath, kind, apiDirName, version string) error {
 	buf := &bytes.Buffer{}
-	if err := renderHandlerFile(buf, repoPath, kind, apiDirName, version); err != nil {
+	if err := renderGenericFile(buf, "stub/handler.go", handlerTmpl, tmplData{
+		OperatorSDKImport: sdkImport,
+		RepoPath:          repoPath,
+		Kind:              kind,
+		APIDirName:        apiDirName,
+		Version:           version,
+	}); err != nil {
 		return err
 	}
 	return writeFileAndPrint(filepath.Join(stubDir, handler), buf.Bytes(), defaultFileMode)
 }
 
+type tmplData struct {
```

Comment:
Most of these fields are pretty self-explanatory. But for fields like `KindSingular` and `KindPlural` could you add a comment please

Reply:
@hasbro17 I wasn't here when the interfaces were designed, can you provide some insight as to why we have 3 versions of Kind?

Reply:
Cool, just updated, should be ready to merge after Travis finishes